Sema: Restrict use of SPI operators to files importing the corresponding SPI group

xymus opened the pull request:
Ensure that calls to SPI operators are only accepted in source files importing the corresponding SPI group. This applies the same logic to operators as we do for functions, patching a hole in type-checking. SPI operators were already correctly detected when used in inlinable code and printed as expected in public module interfaces only.

rdar://137713966

xymus committed:
Sema: Restrict use of SPI operators to files importing the SPI group

Ensure that calls to SPI operators are only accepted in source files
importing the corresponding SPI group. This applies the same logic to
operators as we do for functions.

rdar://137713966

xymus committed:
Sema: Test that SPI operators are only printed in the public swiftinterface

This was already working as expected. Adding a test to ensure it doesn't
regress.

xymus committed:
Sema: Test that we can't use SPI operators in inlinable code

xymus commented:
@swift-ci Please smoke test

tshortli commented on test/SPI/private_swiftinterface.swift:
```diff
@@ -29,16 +29,16 @@
 // RUN: %target-swift-frontend -typecheck %s -emit-module-interface-path %t/Main.swiftinterface -emit-private-module-interface-path %t/Main.private.swiftinterface -enable-library-evolution -swift-version 5 -I %t -module-name Main
-// RUN: %FileCheck -check-prefix=CHECK-PUBLIC %s < %t/Main.swiftinterface
-// RUN: %FileCheck -check-prefix=CHECK-PRIVATE %s < %t/Main.private.swiftinterface
+// RUN: cat %t/Main.swiftinterface | %FileCheck -check-prefix=CHECK-PUBLIC %s
```
Why use `cat` instead of `<` here, out of curiosity? FWIW, switching to `--input-file` instead would improve debugging somewhat (that way FileCheck will print the path to the file when there are mismatches)

xymus replied:
That was it, I prefer that style now to see the name of the file in the outputs. I'll take a look at `--input-file`, maybe it will be worth switching style again.

tshortli commented on lib/AST/Decl.cpp:
```diff
@@ -4832,9 +4832,17 @@ static bool checkAccess(const DeclContext *useDC, const ValueDecl *VD,
   case AccessLevel::Public:
-  case AccessLevel::Open:
+  case AccessLevel::Open: {
+    if (VD->isOperator() && VD->isSPI()) {
```
It might be worth a comment explaining why operators need to be special cased here. I'm having trouble following the surrounding logic and why the existing centralized logic does not apply to them.

xymus committed:
Sema: Gate SPI operator fix behind EnforceSPIOperatorGroup

This new check is source breaking. Let's gate it behind an opt-in flag
for now and enable it by default on a new language mode or similar.

xymus commented:
@swift-ci Please smoke test

xymus commented:
@swift-ci Please smoke test macOS
